File: kanal-core/src/examples/java/io/github/bigironcheems/kanal/examples/TypedBusExample.java
package io.github.bigironcheems.kanal.examples;

import io.github.bigironcheems.kanal.*;

/**
 * Demonstrates {@link TypedEventBus} from Java.
 * <p>
 * A typed bus restricts the domain to a specific event hierarchy, providing
 * compile-time safety that only compatible event types are posted or subscribed.
 * It delegates to an underlying {@link EventBus} for dispatch.
 */
public class TypedBusExample {

    // Event hierarchy

    interface NetworkEvent extends Event {
    }

    static class PacketReceived implements NetworkEvent {
        final byte[] bytes;

        PacketReceived(byte[] bytes) {
            this.bytes = bytes;
        }
    }

    static class ConnectionLost implements NetworkEvent {
        final String reason;

        ConnectionLost(String reason) {
            this.reason = reason;
        }
    }

    // Typed bus: subscribe, post, subscribeAll

    static void typedBus() {
        EventBus underlying = EventBus.create();
        TypedEventBus<NetworkEvent> networkBus =
                TypedEventBusFactory.typed(underlying, NetworkEvent.class);

        networkBus.subscribe(PacketReceived.class, Priority.NORMAL,
                e -> IO.println("Received " + e.bytes.length + " bytes"));
        networkBus.subscribe(ConnectionLost.class, Priority.NORMAL,
                e -> IO.println("Lost: " + e.reason));

        networkBus.post(new PacketReceived(new byte[]{1, 2, 3}));
        networkBus.post(new ConnectionLost("timeout"));
    }

    // Wildcard on a typed bus: fires for every event in the domain

    static void typedBusWildcard() {
        EventBus underlying = EventBus.create();
        TypedEventBus<NetworkEvent> networkBus =
                TypedEventBusFactory.typed(underlying, NetworkEvent.class);

        networkBus.subscribe(PacketReceived.class, Priority.NORMAL,
                e -> IO.println("Received " + e.bytes.length + " bytes"));

        networkBus.subscribeAll(Priority.LOWEST,
                e -> IO.println("[log] " + e.getClass().getSimpleName()));

        networkBus.post(new PacketReceived(new byte[0]));
        // Output:
        //   Received 0 bytes
        //   [log] PacketReceived
    }

    // getDelegate: access the underlying bus if needed

    static void typedBusDelegate() {
        EventBus underlying = EventBus.create();
        TypedEventBus<NetworkEvent> networkBus =
                TypedEventBusFactory.typed(underlying, NetworkEvent.class);

        IO.println("Same underlying bus: " + (networkBus.getDelegate() == underlying)); // true
    }

    public static void main(String[] args) {
        IO.println("=== Typed bus ===");
        typedBus();
        IO.println("\n=== Typed bus wildcard ===");
        typedBusWildcard();
        IO.println("\n=== Typed bus delegate ===");
        typedBusDelegate();
    }
}

File: src/examples/kotlin/io/github/bigironcheems/kanal/examples/TypedBusExample.kt
package io.github.bigironcheems.kanal.examples

import io.github.bigironcheems.kanal.*

// Typed bus
//
// TypedEventBus<E> restricts post and subscribe to subtypes of E at compile
// time. Works best with sealed hierarchies; sealed interface prevents external
// types from satisfying T : E, so the constraint cannot be circumvented.

sealed interface NetworkEvent : Event
class PacketReceived(val bytes: ByteArray) : NetworkEvent
class ConnectionLost(val reason: String) : NetworkEvent

sealed interface UIEvent : Event
class ButtonClickEvent(val id: String) : UIEvent

fun typedBusUsage() {
    val underlying = EventBus()

    // Two domain-scoped views over the same underlying bus
    val networkBus = underlying.typed<NetworkEvent>()
    val uiBus = underlying.typed<UIEvent>()

    networkBus.subscribe<PacketReceived> { e ->
        println("Received ${e.bytes.size} bytes")
    }
    networkBus.subscribe<ConnectionLost> { e ->
        println("Lost connection: ${e.reason}")
    }

    uiBus.subscribe<ButtonClickEvent> { e ->
        println("Button clicked: ${e.id}")
    }

    networkBus.post(PacketReceived(byteArrayOf(1, 2, 3)))
    networkBus.post(ConnectionLost("timeout"))
    uiBus.post(ButtonClickEvent("submit"))

    // networkBus.post(ButtonClickEvent("submit"))  // compile error; ButtonClickEvent is not a NetworkEvent
}

// Typed bus; delegate access

fun typedBusDelegate() {
    val underlying = EventBus()
    val networkBus = underlying.typed<NetworkEvent>()

    networkBus.subscribe<PacketReceived> { }

    // Operations not on the TypedEventBus interface go through delegate
    networkBus.delegate.unsubscribeAll()
    println(networkBus.isListening<PacketReceived>()) // false
}

// Typed bus; unsubscribe annotation-based handler

class NetworkHandler {
    @Subscribe
    fun onPacket(e: PacketReceived) {
        println("handling packet")
    }
}

fun typedBusAnnotationUnsubscribe() {
    val networkBus = EventBus().typed<NetworkEvent>()
    val handler = NetworkHandler()

    networkBus.subscribe(handler)
    println("Before unsubscribe:")
    networkBus.post(PacketReceived(byteArrayOf()))  // fires

    networkBus.unsubscribe(handler)
    println("After unsubscribe:")
    networkBus.post(PacketReceived(byteArrayOf()))  // does not fire
    println("(no output above; handler removed)")
}

fun main() {
    println("=== Typed bus usage ===")
    typedBusUsage()
    println("\n=== Typed bus delegate ===")
    typedBusDelegate()
    println("\n=== Typed bus annotation unsubscribe ===")
    typedBusAnnotationUnsubscribe()
}
